### DISCUSSION TITLE
Conditional variables with generic not getting narrowed

### DISCUSSION THREAD

**Discussion by kossyrev-bg** [microsoft/pyright#6445] [Q&A] [answered]
I'm trying to figure out where my intuition is wrong. I have a class whose function returns different types depending on how it's initialized. I've tried to annotate it with generics to cover this, but it doesn't seem to narrow the return type when I pass in a specific type in the initialization.
```
from typing import Generic, TypeVar, Awaitable

T = TypeVar("T")

class Async:
    def fn(self, returnable: T) -> Awaitable[T]:
        ...

class Sync:
    def fn(self, returnable: T) -> T:
        ...

Client = TypeVar("Client", Async, Sync)

class MyClass(Generic[Client]):
    def __init__(self, client: Client):
        self._client = client
        
    def one_method(self):
        return self._client.fn(7)
    
    
x = MyClass(Async())  # type of X is correctly made MyClass[Async]

z = x._client.fn(7)  # This returns the expected type of Awaitable[int]
r = x.one_method()  # This returns an unexpected type of Awaitable[int]* | int*
```
I'm confused on why `r` is different from `z` in this case.

**Comment by erictraut** [comment, marked as answer]
In your example, the method `one_method` has an inferred return type of `int* | Awaitable[int]*`. The `*` indicates that the type is conditioned on the value of the type variable `Client`. If `Client` has the value `Sync`, then `int` is applicable, and when it has the value `Async`, then `Awaitable[int]` is applicalbe.

When binding `one_method` to an object of type `MyClass[Async]`, pyright should eliminate the subtype of the union that no longer applies (in this case, `int*`). So this appears to be a bug. I'll create a new bug report out of the discussion.

**Reply by erictraut** [reply]
For reference, here's the [bug report](https://github.com/microsoft/pyright/issues/6446).
